src/compiler/msvc.rs

```diff
@@ -513,7 +513,10 @@ pub fn parse_arguments(
                     Argument::Raw(ref val) => {
                         if input_arg.is_some() {
                             // Can't cache compilations with multiple inputs.
-                            cannot_cache!("multiple input files");
+                            cannot_cache!(
+                                "multiple input files",
+                                val.clone().into_string().unwrap()
```

Comment:
```suggestion
                                val.to_string().or_else(|_e| { /* raise invalid args */})?
```
`unwrap`s are generally a bad idea in production code, use `expect` with a proof statement if it's really necessary.

Reply:
This suggestion produces the following two errors:
```rust
error[E0599]: the method `to_string` exists for reference `&OsString`, but its trait bounds were not satisfied
   --> src/compiler/msvc.rs:518:37
    |
518 |   ...                   val.to_string().or_else(|_e| { /* raise invalid args */})?
    |                             ^^^^^^^^^
    |
    = note: the following trait bounds were not satisfied:
            `OsString: std::fmt::Display`
            which is required by `OsString: ToString`
            `&OsString: std::fmt::Display`
            which is required by `&OsString: ToString`
            `OsStr: std::fmt::Display`
            which is required by `OsStr: ToString`

error[E0277]: the `?` operator can only be used in a function that returns `Result` or `Option` (or another type that implements `FromResidual`)
   --> src/compiler/msvc.rs:518:90
    |
452 | / pub fn parse_arguments(
453 | |     arguments: &[OsString],
454 | |     cwd: &Path,
455 | |     is_clang: bool,
...   |
518 | |                                 val.to_string().or_else(|_e| { /* raise invalid args */})?
    | |                                                                                          ^ cannot use the `?` operator in a function that returns `compiler::compiler::CompilerArguments<c::ParsedArguments
>`
...   |
706 | |     })
707 | | }
    | |_- this function should return `Result` or `Option` to accept `?`
    |
    = help: the trait `FromResidual<_>` is not implemented for `compiler::compiler::CompilerArguments<c::ParsedArguments>`

Some errors have detailed explanations: E0277, E0599.
For more information about an error, try `rustc --explain E0277`.
error: could not compile `sccache` due to 2 previous errors
```

Reply:
Mea culpa, I thought it was a `str` rather than an `OsString`/`OsStr`. In that case using the iterator and converting and joining as part of a `fold(||{ TBD })` closure are preferred over the currently proposed impl. (though functional equivalent, this one is a bit more semantic).

Reply:
Pushed a fix replacing the usage of `unwrap` with `fold` over the option iterator. Not sure if this is what you had in mind and if my implementation of it is correct, but it does build and the tests pass as far as I can tell.

I would be much obliged if you could you please point me to any discussion detailing on why this is the preferred implementation? I arrived at the `unwrap` implementation by following rust language reference. Folding an option iterator didn't even come to my mind. Is it pure semantics or is there, for example, some performance gain to it as well?